pkg/autoscaler/statforwarder/forwarder.go

```diff
@@ -0,0 +1,231 @@
+	"knative.dev/pkg/reconciler"
+	"knative.dev/pkg/system"
+)
+
+const (
+	// The port on which autoscaler WebSocket server listens.
+	autoscalerPort     = 8080
+	autoscalerPortName = "http"
+)
+
+// Forwarder does the following things:
+// 1. Watches the change of Leases for Autoscaler buckets. Stores the
+//    Lease -> IP mapping.
+// 2. Creates/updates the corresponding K8S Service and Endpoints.
+// 3. Can be used to forward the metrics falling in a bucket based on
+//    the holder IP. (This is a TODO)
+type Forwarder struct {
+	selfIP          string
+	logger          *zap.SugaredLogger
+	kc              kubernetes.Interface
+	endpointsLister corev1listers.EndpointsLister
+	// bs is the BucketSet including all Autoscaler buckets.
+	bs *hash.BucketSet
+	// Lock for leaseHolders.
+	leaseLock sync.RWMutex
+	// leaseHolders stores the Lease -> IP relationships.
+	leaseHolders map[string]string
+}
+
+// New creates a new Forwarder.
+func New(ctx context.Context, logger *zap.SugaredLogger, kc kubernetes.Interface, selfIP string, bs *hash.BucketSet) *Forwarder {
+	ns := system.Namespace()
+	bkts := bs.Buckets()
+	endpointsInformer := endpointsinformer.Get(ctx)
+	f := Forwarder{
+		selfIP:          selfIP,
+		logger:          logger,
+		kc:              kc,
+		endpointsLister: endpointsInformer.Lister(),
+		bs:              bs,
+		leaseHolders:    make(map[string]string, len(bkts)),
+	}
+
+	leaseInformer := leaseinformer.Get(ctx)
+	leaseInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: reconciler.NamespaceFilterFunc(ns),
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc:    f.leaseUpdated,
+			UpdateFunc: controller.PassNew(f.leaseUpdated),
+			// TODO(yanweiguo): Set up DeleteFunc.
+		},
+	})
+
+	return &f
+}
+
+func (f *Forwarder) leaseUpdated(obj interface{}) {
+	l := obj.(*coordinationv1.Lease)
+	ns, n := l.Namespace, l.Name
+
+	if !f.bs.HasBucket(n) {
+		// Skip this Lease as it is not for Autoscaler.
+		return
+	}
+
+	if l.Spec.HolderIdentity == nil || *l.Spec.HolderIdentity == "" {
+		f.logger.Debugf("HolderIdentity not found for Lease %s/%s", ns, n)
+		return
+	}
+
+	identity := *l.Spec.HolderIdentity
+	if h, ok := f.getHolder(n); ok && h == identity {
+		// Already up-to-date.
+		return
+	}
+
+	f.setHolder(n, identity)
+
+	if identity != f.selfIP {
+		// Skip creating/updating Service and Endpoints as not the holder.
+		return
+	}
+
+	if err := f.createService(ns, n); err != nil {
+		f.logger.Errorf("Failed to create Service for Lease %s/%s: %v", ns, n, err)
+		return
+	}
+	f.logger.Infof("Created Service for Lease %s/%s", ns, n)
+
+	if err := f.createOrUpdateEndpoints(ns, n); err != nil {
+		f.logger.Errorf("Failed to create Endpoints for Lease %s/%s: %v", ns, n, err)
+		return
+	}
+	f.logger.Infof("Created Endpoints for Lease %s/%s", ns, n)
+}
+
+func (f *Forwarder) createService(ns, n string) error {
+	return wait.ExponentialBackoff(retry.DefaultRetry, func() (bool, error) {
```

Comment:
I'm a bit worried about the failure cases here. Right now iiuc we try for a bit (50ms-ish, I think?), but if we fail we just log that fact and end up silently (from an API PoV) in an invalid state. Is there some way we can surface this error so an operator can see this apart from looking for errors in the log (or is there one already I just didn't see?). Even better is there any way we can recover here?

Reply:
If there's a way to recover, I'd like to do. Do you have suggestions?

Reply:
Im sort of wondering if we should have a work queue here so we can re-enqueue this a while later, but I don't know if it's overkill. Another option maybe is to fail a readiness check, since we're genuinely in a bad state. I'm not sure though.. I just think being silently and unrecoverably in a bad state seems wrong.. any ideas @vagababov?

Reply:
I can support a short poll, rather than exp backoff followed by a fatal or liveness check failure.
If we can't talk to API server, we should not be holding leases.

Reply:
Added a TODO.

In activator, we also log in a AddFunc: https://github.com/knative/serving/blob/9c863c8d79a863df8500d7029468291004fbac52/pkg/activator/net/throttler.go#L557. Do we need to improve there as well?

Reply:
Also changed to PollImmediate